### PR TITLE
Fix the bug for restoring data with duplicated page id

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/3594

Problem Summary:
https://github.com/pingcap/tiflash/issues/4872
https://github.com/pingcap/tiflash/issues/4849
We fix those two bugs with previous PRs, but those changes introduce new misbehavior for TiFlash. Find a better way to fix those bugs.

### What is changed and how it works?

Fix reusing same page_id for different kind by:
* Throw exception with error code == PS_DIR_APPLY_INVALID_STATUS when `VersionedPageEntries:: createNewEntry/createNewExternal/createNewRef/createNewExternal/createNewRef` run into invali state
* When restoring edit run into exception with error code == PS_DIR_APPLY_INVALID_STATUS, then call GC on mvcc_directory  entries and run apply edit again

Other changes:
* When restoring the storage pool for DeltaTree, get the max id for this ns_id by `PageStorage->getMaxId`
* `PageStorage->getMaxId` now ignore the invisible id
* Instead of `meta_max_id`, use `PageEntry.isValid()` to check whether page 1 is existed or not https://github.com/pingcap/tiflash/blob/3e46ac2a411560503abba977a37a5caa495e9bd2/dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp#L246-L247
* Enable some tests that are disabled in the previous changes

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/tests/gtest_dm_storage_delta_merge.cpp [hunk 1]
```diff
@@ -40,6 +40,7 @@
 
 #include <limits>
 
+#include "Storages/Transaction/TiDB.h"
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/tests/gtest_dm_storage_delta_merge.cpp:43]:
```suggestion
#include <Storages/Transaction/TiDB.h>
```

File: dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp [hunk 2]
```diff
@@ -101,80 +107,94 @@ PageDirectoryPtr PageDirectoryFactory::createFromEdit(String storage_name, FileP
...
         max_applied_page_id = std::max(r.page_id, max_applied_page_id);
 
-        auto [iter, created] = mvcc_table_directory.insert(std::make_pair(r.page_id, nullptr));
-        if (created)
+        // We can not avoid page id from being reused under some corner situation. Try to do gcInMemEntries
+        // and apply again to resolve the error.
+        if (bool ok = applyRecord(dir, r, /*throw_on_error*/ false); unlikely(!ok))
         {
-            iter->second = std::make_shared<VersionedPageEntries>();
+            dir->gcInMemEntries();
```
Comment (hehechen) [inline dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:120]: Will it cause restore performance degradation if there are many edit entries?

File: tests/delta-merge-test/raft/schema/partition_table_restart.test [hunk 3]
```diff
@@ -35,8 +35,6 @@
 # schema syncer guarantees logical table creation at last, so there won't be cases that logical table exists whereas physical table not.
 => drop table default.test
 => DBGInvoke __reset_schemas()
-# remove obsolete entry left by previous dropped table
-=> DBGInvoke __gc_global_storage_pool()
```
Comment (JaySon-Huang) [inline tests/delta-merge-test/raft/schema/partition_table_restart.test:39]: Seems we still need this workaround, will investigate why

File: dbms/src/Storages/Page/V3/PageDirectory.cpp [hunk 4]
```diff
@@ -904,10 +939,31 @@ PageId PageDirectory::getMaxId(NamespaceId ns_id) const
...
+            }
+
+            // Current entry/ref/external is deleted and there are no entries before it.
+            if (iter == mvcc_table_directory.begin())
+            {
+                break;
+            }
+
+            iter--;
+        } while (true);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/PageDirectory.cpp:964]: These changes revert https://github.com/pingcap/tiflash/pull/4845 but use `isVisible` to check whether we can return the id in getMaxId /cc @jiaqizho @lidezhu

File: dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp [hunk 5]
```diff
@@ -243,7 +243,8 @@ DeltaMergeStore::DeltaMergeStore(Context & db_context,
     try
     {
         page_storage_run_mode = storage_pool->restore(); // restore from disk
-        if (!storage_pool->maxMetaPageId())
+        if (const auto entry1 = storage_pool->metaReader()->getPageEntry(1);
```
Comment (flowbehappy) [inline dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp:246]: Let's give `1` a name.

Comment (lidezhu) [reply dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp:246]: It's already has a name `DELTA_MERGE_FIRST_SEGMENT_ID`.

File: dbms/src/Storages/Page/V3/PageDirectory.cpp [hunk 6]
```diff
@@ -375,6 +380,41 @@ std::optional<PageEntryV3> VersionedPageEntries::getLastEntry() const
     return std::nullopt;
 }
 
+// Returns true when **this id** is visible (not exist or already marked as deleted) by `seq`.
```
Comment (flowbehappy) [inline dbms/src/Storages/Page/V3/PageDirectory.cpp:383]: "not exist or already marked as deleted" seems odd.

File: dbms/src/Interpreters/Context.cpp [hunk 7]
```diff
@@ -1650,8 +1650,10 @@ bool Context::initializeGlobalStoragePoolIfNeed(const PathPool & path_pool)
     auto lock = getLock();
     if (shared->global_storage_pool)
     {
-        // GlobalStoragePool may be initialized many times in some test cases for restore.
-        LOG_WARNING(shared->log, "GlobalStoragePool has already been initialized.");
+        // Can't init GlobalStoragePool twice.
+        // Because we won't remove the gc task in BackGroundPool
```
Comment (jiaqizho) [inline dbms/src/Interpreters/Context.cpp:1654]: Change back this comment pls. Also need change back to warning.

File: dbms/src/Storages/DeltaMerge/StoragePool.cpp [hunk 8]
```diff
@@ -375,18 +367,18 @@ PageStorageRunMode StoragePool::restore()
             data_storage_writer = std::make_shared<PageWriter>(run_mode, /*storage_v2_*/ nullptr, data_storage_v3);
             meta_storage_writer = std::make_shared<PageWriter>(run_mode, /*storage_v2_*/ nullptr, meta_storage_v3);
 
-            max_log_page_id = global_storage_pool->getLogMaxId(ns_id);
-            max_data_page_id = global_storage_pool->getDataMaxId(ns_id);
-            max_meta_page_id = global_storage_pool->getMetaMaxId(ns_id);
+            max_log_page_id = global_storage_pool->log_storage->getMaxId(ns_id);
```
Comment (jiaqizho) [inline dbms/src/Storages/DeltaMerge/StoragePool.cpp:370]: Use `log_pagestorage_v3` to replace `global_storage_pool->log_storage` pls.
also dito meta/data and below.

File: dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp (hunk 2 above)
Comment (jiaqizho) [inline dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:120]: If there is too much multiplexing, will gcInMemEntries be triggered too many times ?
Maybe we can add an interface name `gcInMemEntries(page_id)`, or add a TODO here ?

File: dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp [hunk 9]
```diff
@@ -101,80 +107,94 @@ PageDirectoryPtr PageDirectoryFactory::createFromEdit(String storage_name, FileP
...
+            applyRecord(dir, r, /*throw_on_error*/ true);
+            LOG_FMT_INFO(DB::Logger::get("PageDirectoryFactory"), "resolve from error status done, continue to restore");
```
Comment (jiaqizho) [inline dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:122]: `DB::Logger::get("PageDirectoryFactory")` add a log in PageDirectoryFactory to replace it? cause it used many place

File: dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp [hunk 10]
```diff
@@ -101,80 +107,94 @@ PageDirectoryPtr PageDirectoryFactory::createFromEdit(String storage_name, FileP
...
+    }
+    catch (DB::Exception & e)
+    {
+        e.addMessage(fmt::format(" [type={}] [page_id={}] [ver={}]", r.type, r.page_id, restored_version));
+        if (throw_on_error || e.code() != ErrorCodes::PS_DIR_APPLY_INVALID_STATUS)
         {
-            e.addMessage(fmt::format(" [type={}] [page_id={}] [ver={}]", r.type, r.page_id, restored_version));
             throw e;
         }
+        LOG_FMT_INFO(DB::Logger::get("PageDirectoryFactory"), "try to resolve error during restore: {}", e.message());
```
Comment (jiaqizho) [inline dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:194]: Can we change this log into warning?

File: dbms/src/Interpreters/Context.cpp (hunk 7 above)
Comment (JaySon-Huang) [reply dbms/src/Interpreters/Context.cpp:1654]: We do NOT re-init GlobalStoragePool even in test cases now.

File: dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp (hunk 2 above)
Comment (JaySon-Huang) [reply dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:120]:
```
PUT_EXT 100
REF 150 -> 100
DEL 100
REF 100 -> 40  <- exception with PS_DIR_APPLY_INVALID_STATUS thrown here
```

We can not only run GC on id==100 to get things right.

File: dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp (hunk 9 above)
Comment (JaySon-Huang) [reply dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp:122]: Now we only create it when *failover* does happen.
Just a small optimization to reduce the size of global `_pLoggerMap` in Poco. If it does not occur, we won't call `Poco::Logger::get` and we won't add a new element to the global `_pLoggerMap`. 

https://github.com/pingcap/poco/blob/e411ea34492bf26e76b4e4a03ff5813a0c779240/Foundation/src/Logger.cpp#L475-L480

File: dbms/src/Storages/DeltaMerge/StoragePool.cpp (hunk 8 above)
Comment (JaySon-Huang) [reply dbms/src/Storages/DeltaMerge/StoragePool.cpp]: Fixed, please check again

**Review by lidezhu — APPROVED** [review]
LGTM

File: dbms/src/Interpreters/Context.cpp (hunk 7 above)
Comment (jiaqizho) [reply dbms/src/Interpreters/Context.cpp:1654]: okay, but the Notes, it's not right. need updated

**Review by jiaqizho — APPROVED** [review]
LGTM, But still need to change the notes(that reason is not right). https://github.com/pingcap/tiflash/blob/f2c47c9a98a40b2c832219f100040e418a7f1e74/dbms/src/Interpreters/Context.cpp#L1654-L1655

**Pull request merged**